### DISCUSSION TITLE
Seeking Guidance on Creating a Mapping from Grain ID to Container_Type Element ID

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#25807] [Q&A Modules: Phase field] [answered]
Dear @permcody and @laagesen,

I hope this message finds you well. I have a question regarding a specific implementation in my project, and I would greatly appreciate your expertise and guidance on this matter.

In my current project, I'm working on creating a mapping structure that associates Grain IDs with container_type Element IDs. This mapping is similar to the existing '_entity_var_to_features' structure, and it is a crucial component for the functionality I am developing.

However, I've encountered a challenge related to the 'FeatureData' structure. I've noticed that the data member 'container_type _local_ids' is only maintained on the local processor. Therefore, when executing parallel computations, I do not have access to all the element IDs contained within 'feature[i]._local_ids'. This limitation has brought up the need to introduce a new data member, which I'm thinking of naming 'container_type _global_ids' within the 'FeatureData' structure. This new data member would be used to store all the Element IDs associated with a particular feature or grain.

My question is twofold:

1. Is it possible to add a new data member, 'container_type _global_ids,' to the 'FeatureData' structure for storing all the Element IDs associated with a feature? If so, what would be the most efficient way to implement this addition? Of course, I also know that this setting will require huge data to be transmitted between ranks and cause greater memory consumption.

2. How can I effectively merge the '_local_ids' from various processors into a single '_global_ids' structure, ensuring that I have a comprehensive list of all the Element IDs for each feature, regardless of the processor?

I would greatly appreciate any insights, advice, or examples you can provide to help me address this challenge effectively. Your expertise in this matter is invaluable, and your guidance will significantly impact the success of my project.

Thank you in advance for your assistance, and I look forward to your response.

Best regards,

wei peng

**Comment by GiudGiud** [comment]
Hello

For 2)
_local_ids is a std::vector<dof_id_type>

these are actually very easy (remains expensive if large) to expand on all processes using TIMPI. 
You simply copy it into your new global container, then use the allgather:
```
    _global_ids = _local_ids;
    _communicator.allgather(_global_ids, /* identical buffer lengths = */ false);
```
you just need to make sure this is called right before you need the information

Guillaume

**Reply by PengWei97** [reply]
Hello @GiudGiud,

Thank you for your previous guidance. I greatly appreciate your help. I spent the entire day working on implementing your suggestions, but unfortunately, the issue is not fully resolved.

Following your advice, I initially created a `std::vector<std::vector<dof_id_type>> _grain_id_to_entity_ids_maps;` to represent `_grain_id_to_entity_ids_maps[grain._id]` as a vector of entity ids corresponding to `grain._id`. I temporarily added the following code to `GrainTracker::updateFieldInfo()`:

```c++
_grain_id_to_entity_ids_maps.resize(_feature_sets.size());

for (const auto & grain : _feature_sets)
{
    std::cout << "grain id: " << grain._id << std::endl;
    for (const auto & entity : grain._local_ids) 
    {
        std::vector<dof_id_type> & entity_ids = _grain_id_to_entity_ids_maps[grain._id];
        if (std::find(entity_ids.begin(), entity_ids.end(), entity) == entity_ids.end())
            entity_ids.push_back(entity);
    }
}

_communicator.allgather(_grain_id_to_entity_ids_maps);

for (unsigned int i = 0; i < _grain_id_to_entity_ids_maps.size(); ++i)
    std::cout << "grain id after allgather: " << i << std::endl;
```

However, I encountered an unexpected issue. When running the code in single-core mode, everything behaves as expected, and the size of `_grain_id_to_entity_ids_maps` is 6. But when running it on two cores, the size becomes 14, which is problematic. I have included a snippet of the output results.
![image](https://github.com/idaholab/moose/assets/56183775/4d338d06-d06a-43ad-b6ac-e09d77e156cb)

Best regards,

wei

**Reply by PengWei97** [reply]
Meanwhile, I believe this inconsistency is due to the `allgather` operation concatenating the `_grain_id_to_entity_ids_maps` from different rank without properly merging the individual `_grain_id_to_entity_ids_maps[grain._id]`. To address this, I attempted to create a std::map<unsigned int, std::vector<dof_id_type >>  type named `_var_index_maps, but I encountered a compilation error during the _communicator.allgather(_grain_id_to_entity_ids_maps) operation.
```bash 
/home/pw-moose/projects/moose/modules/phase_field/src/postprocessors/GrainTracker.C: In member function 'virtual void GrainTracker::updateFieldInfo()':
/home/pw-moose/projects/moose/modules/phase_field/src/postprocessors/GrainTracker.C:1592:55: error: no matching function for call to 'libMesh::Parallel::Communicator::allgather(std::map<unsigned int, std::vector<long unsigned int> >&) const'
 1592 |   _communicator.allgather(_grain_id_to_entity_ids_maps);
```

Finally, I feel that there should be a step in GrainTracker or FeatureFloodCount to merge the _feature_sets or other data members on each rank into the primary rank to form a global _feature_sets or others. But unfortunately, I did not find the relevant code in `communicateAndMerge()`, `broadcastAndUpdateGrainData()` and `updateFieldInfo()`. This may be because I don't understand GrainTracker very well, but I'm sure the code should be there.

In general, I would greatly appreciate any further guidance or recommendations to address this issue. Your assistance is invaluable.

**Reply by GiudGiud** [reply]
Hello

If the gathering operation is working well, then you may perform the merging / "unique" operation on each rank considering all the data that was gathered. It s not a scalable operation but the communication step is not either

Did you consider using a set since you want unique data? I m not sure what you need is implemented for _comm.allgather for a set

Guillaume

**Reply by PengWei97** [reply]
Hello,

I believe the following schematic can help illustrate the dataset I'm trying to merge, which is represented by `_grain_id_to_entity_ids_maps`.
<img width="288" alt="image" src="https://github.com/idaholab/moose/assets/56183775/5a405486-a49b-40fc-8ffd-16b164017f96">

As shown in the diagram, let's assume that in a parallel computation, the computational domain containing Gr0 and Gr1 is divided into two parts by the red dashed line, with one part being calculated on Rank 0 and the other on Rank 1. Elements or entities belonging to Gr0 are represented by white dots, and there are three of them on both Rank 0 and Rank 1, with their element IDs stored in `_feature_sets[0]._local_ids`.

Now, if only extract the mapping of the element id corresponding to gr0, the challenge is how to ensure that `_grain_id_to_entity_ids_maps[0] = [0, 1, 2, 3, 4, 5]` instead of `_grain_id_to_entity_ids_maps[0] = [0, 1, 2]` and `_grain_id_to_entity_ids_maps[1] = [3, 4, 5]`.

Thank you,

Wei Peng

**Reply by GiudGiud** [reply]
You should try to do the gather operation on the map

If that does not work then do a reduction on a keys and a value vector

**Reply by PengWei97** [reply]
Hi @GiudGiud, thank you very much for your further guidance. But I think you may be optimistic that I have the relevant understanding of MPI in moose and MPI itself. In fact, I don’t know much about MPI. Of course, I also tried to learn MPI, but it seemed that what I saw was not the same as what I need to use in moose.

So, to be honest,
> If that does not work then do a reduction on a keys and a value vector

I don't know how to do it. In addition, I also look forward to your sharing of learning materials about the use of MPI in moose.

I set the data type to `std::map<unsigned int, std::vector<dof_id_type>> _grain_id_to_entity_ids_maps`, and use the following code to run, but unfortunately, of course, when adding the red box selection code , still showing the same error mentioned above.
![7bf28417e5210b7e3796c637bff2347](https://github.com/idaholab/moose/assets/56183775/ae809e6b-59ad-4a7c-9998-64f53e50f857)

Thank you,

Wei Peng

**Reply by GiudGiud** [reply]
Looking at the TIMPI documentation there is no overload for maps of allgather, so that s why you get the error
https://mooseframework.inl.gov/docs/doxygen/timpi/classTIMPI_1_1Communicator.html#af7a4310c8ff9eff4deee15382b4a1a62

> If that does not work then do a reduction on a keys and a value vector

You can copy the grain ids and entity in separate vectors then gather those.
Once you have both on a rank you can work to merge the data 

If this is for postprocessing purposes, I reckon you do not need allgather, you can get away with a simple gather to rank 0.

**Reply by PengWei97** [reply]
I carefully read the link you shared with me and looked at the four defined gather templates. Unfortunately, I did not find that gather can handle such things as gather(0, std::vector<unsigned int> & send_data, std::vector<std::vector<unsigned int>> & recv_data). In other words, the gather operation can only merge the std::vector<unsigned int> data on each rank into std::vector<unsigned int> data, and cannot create a dimension based on different ranks to form std::vector. <std::vector<unsigned int>> data_ranks;

> You can copy the grain ids and entity in separate vectors then gather those.

According to your suggestion, if they are stored separately and then transferred to rank 0, then I think the correspondence between grain id -> element ids will be lost during the transfer process.

I believe you already have a good understanding of what I intend to implement, and I would greatly appreciate any additional suggestions or perhaps even some sample code.

wei

**Comment by permcody** [comment]
Sorry I’m late to the conversation. I’d like to understand more of why you
believe you need non-local information on all ranks. If you do need that
information. The best bet would be to gather it all on one rank or at least
a small subset of ranks to perform a calculation. As Guillaume mentioned,
doing an allgather is expensive! This won’t scale to run on a larger 3D
problem.

Instead consider performing a partial calculation on each processor and
then gathering the relevant information on the first rank for final
processing. This what we do in the FeatureVolumeVectorPostprocessor.

You can proceed with doing the unique merge if you’d like but there is a
shortcut. We already have the complete grain information at the end of the
merge stage. We purposely split that up and only send local information to
each rank. You could just skip that step and do a broadcast there instead
of the “scatter” operation that’s there.

**Reply by PengWei97** [reply]
Hi @permcody, thank you very much for your help and guidance many times. In fact, the purpose of collecting each feature_set._local_ids on all ranks is to calculate the average of a certain material property of all elements corresponding to each grain, such as Cauchy stress. After obtaining the average value corresponding to each particle, the model calculation is performed by calling the actual object, as mentioned in Chen Longqing's article AA.

I'm also well aware that doing a full gathering is expensive! At the same time, I still have some questions about the two solutions proposed later, and  I would appreciate your insights.

Based on `FeatureVolumeVectorPostprocessor`, how should the average value of each material parameter on each rank be calculated? That is, how to complete the calculation of \sigma_{n_grain, n_rank, ave} = \sum_{n_elem=1}^{N}\sigma_{n_grain, n_rank, n_elem} for each grain (n_grain), where n_grain represents grain DI, n_rank represents rank ID, and n_elem represents element id. Then collect \sigma_{n_grain, n_rank, ave} on each rank to calculate \sigma_{n_rank, ave}.

Regarding the second suggestion, you mentioned: "We already have the complete grain information at the end of the merge stage. We purposely split that up and only send local information to each rank." Where is this part of the code located? Whether _feature_sets[i]._local_ids contains complete element ids information on primary rank before executing "scattering".

Maybe I'll try the shortcuts first and then try the first optimal solution after the basic functionality is complete.

Best regards,

Wei

**Reply by permcody** [reply, marked as answer]
> Based on FeatureVolumeVectorPostprocessor, how should the average value of each material parameter on each rank be calculated? That is, how to complete the calculation of \sigma_{n_grain, n_rank, ave} = \sum_{n_elem=1}^{N}\sigma_{n_grain, n_rank, n_elem} for each grain (n_grain), where n_grain represents grain DI, n_rank represents rank ID, and n_elem represents element id. Then collect \sigma_{n_grain, n_rank, ave} on each rank to calculate \sigma_{n_rank, ave}.

This is exactly what we do in `FeatureVolumeFectorPostprocessor`. We still do a loop over all local elements on each processor. That's done here: https://github.com/idaholab/moose/blob/bdad35fcdd7aa0bf79eaaa60e3372853c09c9023/modules/phase_field/src/vectorpostprocessors/FeatureVolumeVectorPostprocessor.C#L157 Inside of that loop, we then calculate the quantity of interest using the shape functions representing that particular feature: https://github.com/idaholab/moose/blob/bdad35fcdd7aa0bf79eaaa60e3372853c09c9023/modules/phase_field/src/vectorpostprocessors/FeatureVolumeVectorPostprocessor.C#L233-L234 Finally, those partial quantities are summed across all elements by feature: https://github.com/idaholab/moose/blob/bdad35fcdd7aa0bf79eaaa60e3372853c09c9023/modules/phase_field/src/vectorpostprocessors/FeatureVolumeVectorPostprocessor.C#L175-L180 

In any case, let the individual processors do partial calculations and then sum everything together at the end. This is much more scalable in terms of time and space. All of the information you need for your calculation is available in that VPP already.

**Reply by PengWei97** [reply]
Great, what I want to achieve has been initially solved. Thank you so much for your crucial guidance. Specifically, I implemented it by creating a new derived class `FeatureMatePropVectorPostprocessor`. The key is to introduce the material properties that need to be averaged:
```c++
sum += _JxW[qp] * _coord[qp] * MetaPhysicL::raw_value(_mat_prop_qp[qp]) * (*_coupled_sln[var_index])[qp];
```
Simultaneously, the entire code [FeatureMatePropVectorPostprocessor](https://github.com/PengWei97/qinglong/commit/7f585c8cc38b64b547c2b4eddbd03a86da51ece1) has been uploaded to the remote repository. If you have any additional suggestions or insights, I would greatly appreciate your help.

Best regards,

wei peng
